open-apparel-registry/open-apparel-registry
# Require a token for all API endpoints

## Overview

Adds an authentication requirement to all API endpoints.

- Required for accurately reporting on API usage.
- Allows for disabling accounts that accidentally or maliciously overuse the APIs.

Maintains the ability for the web application to make API requests on behalf of unauthenticated users.

Connects #636

## Demo

<img width="1016" alt="Screen Shot 2019-07-02 at 9 24 56 AM" src="https://user-images.githubusercontent.com/17363/60529368-4a377580-9cab-11e9-85a8-ae8576be716a.png">

<img width="1071" alt="Screen Shot 2019-07-02 at 9 20 21 AM" src="https://user-images.githubusercontent.com/17363/60529297-283df300-9cab-11e9-8cd1-42464dc1bee5.png">

<img width="1072" alt="Screen Shot 2019-07-02 at 9 20 02 AM" src="https://user-images.githubusercontent.com/17363/60529305-2d02a700-9cab-11e9-8550-02b98c2a4646.png">


## Notes

I have added an `OAR_CLIENT_KEY` to the staging tfvars file. There is a checklist item on this PR to make sure the next release issue includes updating the production tfvars file.

## Testing Instructions

### Setup

* `./scripts/manage resetdb`
* `./scripts/manage loadfixtures`
* `./scripts/manage processfixtures`

#### Enable Rollbar in development

* Set the `ROLLBAR_SERVER_SIDE_ACCESS_TOKEN` value in `.env.backend`
* Enable Rollbar in `settings.py`
```diff
diff --git a/src/django/oar/settings.py b/src/django/oar/settings.py
index 10c2f02..bb0a11e 100644
--- a/src/django/oar/settings.py
+++ b/src/django/oar/settings.py
@@ -338,7 +338,7 @@ if GOOGLE_SERVER_SIDE_API_KEY is None:
     raise ImproperlyConfigured(
         'Invalid GOOGLE_SERVER_SIDE_API_KEY provided, must be set')
 
-if not DEBUG:
+if True:
     ROLLBAR = {
         'access_token': os.getenv('ROLLBAR_SERVER_SIDE_ACCESS_TOKEN'),
         'environment': ENVIRONMENT.lower(),
```

### Test

* Start `./scripts/server`.
* Ensure that you are not logged in to the application
* Verify that you can use the application as normal.
* Verify that `curl http://localhost:6543/api/countries/` returns a response.
* Set an `OAR_CLIENT_KEY` env var in `docker-compose.yml` and restart `./scripts/server`

```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index a769732..1f477a4 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -25,7 +25,7 @@ services:
         - DJANGO_SECRET_KEY=secret
         - DJANGO_LOG_LEVEL=INFO
         - AWS_PROFILE=${AWS_PROFILE:-open-apparel-registry}
-        - OAR_CLIENT_KEY=
+        - OAR_CLIENT_KEY=clientkey
     build:
       context: ./src/django
       dockerfile: Dockerfile
```
* Verify that you can still use the application as normal. 
* Verify that `curl http://localhost:6543/api/countries/` now returns an error message and that an "Incorrect client key submitted with API request" occurrence is logged to Rollbar
```javascript
{"detail":"Authentication credentials were not provided."}
```
* Make a request with a client key header
```
curl \
  -H "X-OAR-Client-Key: clientkey" \
  http://localhost:6543/api/countries/
````
* Verify that the same authentication error is returned, but an "Unallowed referring host passed with API request" warning is logged in Rollbar.
* Make a request with a client key and a referrer
```
curl \
  -H "X-OAR-Client-Key: clientkey" \
  -H "Referer: http://localhost:6543" \
  http://localhost:6543/api/countries/
```
* Verify that it is sucessful
* Log in to the web app as `c2@example.com`. 
* Browse http://localhost:6543/profile/2 and generate an API token
* Make an authenticated request with the token and verify that it succeeds.
```
curl \
  -H "Authorization: Token {{API-TOKEN}}" \
  http://localhost:6543/api/countries/
```

## Checklist

- [x] Add a checklist item to the next release issue for adding an `oar_client_key` to  the tfvars.
